Add a lightweight scheduler subsystem

peter-mitsis opened the pull request:
There has been a substantial re-work based on various comments. Instead of being a port of the SOF LL scheduler, this is now merely inspired by it and called the Lightweight Scheduler.

The basic premise is this.  A "scheduling thread" is triggered at regular intervals by a k_timer. During each interval, it runs through its list of registered "tasks" in sequence and calls the associated handler if it is in the executing state. These tasks can be in one of three states: paused (which is also the initial state), execute, and abort.

More information is in the documentation commit.

peter-mitsis commented:
Working on fixing these checks ....

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+LL scheduler from the audio SOF project. At the current time, it only supports
+a single scheduler that readies tasks (aka threads) to run based solely on the
+timer interrupt. The order in which these tasks execute is still handled by
+the classic Zephyr scheduling algorithm which is priority based.
```
```suggestion
the classic Zephyr scheduling algorithm, which is priority-based.
```

kartben replied:
outdated

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+Specifying the Scheduling Domain
+================================
+
+When specifying the scheduling two things must be done. First, the scheduling
```
```suggestion
When specifying the scheduling, two things must be done. First, the scheduling
```

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+task will be scheduled to execute according to the domain with which it is
+registered. That is, when registered with the LL timer domain, it will be
+scheduled to execute every CONFIG_LL_SCHEDULER_TIMER_PERIOD microseconds.
+It is incumbent upon the system designer to register tasks with this that
+do not violate this timing.
```
```suggestion
The system designer must ensure that registered tasks do not violate this timing.
```

kartben replied:
outdated

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+#include <stddef.h>
+#include <stdint.h>
+
+/** \brief Type of scheduler, comes from topology. */
```
the file has a mix of `\brief`, `@brief`, and implicit briefs. I would suggest to use the same approach for the entire file (I think implicit works nicely).

peter-mitsis replied:
Will use @brief moving forward as that appears to be most common in our source.

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
This needs to define a Doxygen group (and have a `@file` annotation too) and be mounted somewhere in the doc hierarchy.

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+	LL_SCHED_COUNT,	/**< indicates number of scheduler types */
+};
+
+struct ll_sched_task;
```
docstring?

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+
+struct ll_sched_task;
+
+struct ll_sched_uuid_entry;
```
docstring?

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+	void *data;			/**< pointer to private data */
+};
+
+enum ll_sched_task_state {
```
missing docstring

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+	uint16_t flags;		/**< custom flags */
+	enum ll_sched_task_state state;	/**< current state */
+	void *data;		/**< custom data passed to all ops */
+	sys_dnode_t      node;	/**< task's node in the scheduler's task list */
```
extra whitespaces in `sys_dnode_t      node`

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+
+/** \brief Task operations. */
+struct ll_sched_task_ops {
+	enum ll_sched_task_state (*run)(void *data);	/**< task's main operation */
+	void (*complete)(void *data);		/**< executed on completion */
+	uint64_t (*get_deadline)(void *data);	/**< returns current deadline */
```
first occurence, with several other incorrectly aligned comments further down: tab size is 8 characters in Zephyr, could you please fix comments alignment accordingly?

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+		       enum ll_sched_task_state (*run)(void *data),
+		       void *data, uint16_t core, uint32_t flags);
+
+struct ll_sched_domain;    /* Type is opaque to user */
```
still need a docstring even if it's opaque

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+
+int ll_sched_scheduler_init(struct ll_sched_domain *domain);
+
+/** See ll_sched_ops::schedule_task_cancel */
```
```suggestion
/** @see ll_sched_ops::schedule_task_cancel */
```

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+	return -ENODEV;
+}
+
+/** See ll_sched_ops::schedule_task */
```
```suggestion
/** @see ll_sched_ops::schedule_task */
```

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+
+struct ll_sched_domain;    /* Type is opaque to user */
+
+struct ll_sched_domain *ll_timer_domain_init(int clk);
```
docstring

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+	return task->ops.run(task->data);
+}
+
+int ll_sched_task_init(struct ll_sched_task *task,
```
docstring

kartben commented on include/zephyr/ll_sched/ll_scheduler.h:
```diff
@@ -0,0 +1,303 @@
+ */
+void ll_sched_free(void *ptr);
+
+static inline enum ll_sched_task_state task_run(struct ll_sched_task *task)
```
docstring

kartben commented on samples/subsys/ll_sched/README.rst:
```diff
@@ -0,0 +1,36 @@
+.. _ll_sched-sample:
+
+Low-Latency (LL) Scheduler
+##########################
```
```suggestion
.. zephyr:code-sample:: ll_sched
   :name: Low Latency (LL) Scheduler
   :relevant-api: HERE_THE_DOXYGEN_GROUP_FOR_NEW_LL_API

   Use the low latency scheduler to schedule a task.
```

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+* :kconfig:option:`CONFIG_LL_SCHEDULER_TIMER`
+* :kconfig:option:`CONFIG_LL_SCHEDULER_TIMER_PERIOD`
+* :kconfig:option:`CONFIG_LL_SCHEDULER_LOG_CYCLE_STATISTICS`
+* :kconfig:option:`CONFIG_LL_SCHEDULER_TIMER_TASK_STACK_SIZE`
```
add API reference, along the lines of

```
API Reference
*************

.. doxygengroup:: HERE_THE_NEW_DOXYGEN_GROUP
```

kartben commented on samples/subsys/ll_sched/README.rst:
```diff
@@ -0,0 +1,36 @@
+
+Overview
+********
+This sample demonstrates the use of the LL scheduler. It ...
```
```suggestion
This sample demonstrates the use of the LL scheduler, and goes through the following steps:
```

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+    struct ll_sched_domain *domain;
+
+    /* Initialize the timer scheduling domain */
+
```
delete

kartben commented on doc/services/ll_sched/index.rst:
```diff
@@ -0,0 +1,157 @@
+    domain = ll_timer_domain_init(0);
+
+    /* Create a scheduler on the current CPU based on this domain */
+
```
delete

kartben reviewed (changes requested):
I can't comment on the technical implementation, but see my review re: documentation. I might have missed a few things, but many of my comments are quite general so I trust you will be able to apply theme everywhere they are relevant, even where I didn't explicitly flag an issue in the form of a review comment. Thanks!

peter-mitsis commented on doc/services/index.rst:
```diff
@@ -15,6 +15,7 @@ OS Services
    formatted_output.rst
    input/index.rst
    ipc/index.rst
+   ll_sched/index.rst
    logging/index.rst
```
Can do. (I see where I erred here ... I was previously going alphabetically by directory name, and not alphabetically by what got generated).

andyross commented:
Catching up on stale review requests.  Not going to stand in the way of this one but... is this really the right thing to do?  I get that this is just code motion from SOF, but (1) it's a LOT of code and (2) it doesn't really... do much?  The original SOF design was thicker, but largely abandoned, and all that's left here is just a bunch of callbacks to be made out of a timer interrupt.

Which is to say: this is just k_timer.  Do we really want a 1600-line subsystem added to the tree just to do what k_timer does already?  Can't that effort be better spent porting SOF internally to use k_timer instead?

nashif commented on subsys/ll_sched/Kconfig:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig LL_SCHEDULER
+	bool "LL Scheduler"
```
call it out here please, LL meaning "Low Level"

peter-mitsis replied:
Done.

nashif commented on subsys/ll_sched/Kconfig:
```diff
@@ -0,0 +1,37 @@
+	default n
+	help
+	  This option enables the low latency (LL) scheduler. It is derived
+	  from the scheduler used in the audio SOF module. Currently, it only
```
no need to have those references to SOF everywhere, just put this in the docs somewhere as reference and that should be enough

peter-mitsis replied:
Done.

peter-mitsis committed:
lw_sched: subsys: Add lightweight scheduler

The lightweight scheduler is a very simple scheduling construct. It
has a dedicated thread that runs at regular intervals to execute a
list of tasks in sequence.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
samples: subsys: Add lightweight scheduler sample

Adds sample application to demo the use of the lightweight
scheduler.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
doc: Documentation for LW scheduler sub-system

Add documentation for the LW scheduler.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
tests: Add lightweight scheduler tests

Adds some basic tests for the LW scheduler.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

kartben commented on include/zephyr/lw_sched/lw_sched.h:
```diff
@@ -0,0 +1,221 @@
+ * @param sched Pointer to the LW scheduler to initialize
+ * @param stack Pointer to the stack used by the LW scheduler's thread
+ * @param stack_size Size of the stack used by the LW scheduler
+ * @param priority Priority of the LW scheduler thread
```
```suggestion
 * @param priority Priority of the LW scheduler's thread
```

kartben reviewed (changes requested):
Looking great -- please see additional comments re: documentation.

kartben commented on include/zephyr/lw_sched/lw_sched.h:
```diff
@@ -0,0 +1,221 @@
+	uint32_t state;                 /**< Task state */
+	uint32_t order;                 /**< Task order */
+	uint32_t new_order;             /**< New task order */
+	uint32_t delay;                 /**< # of intervals to delay */
```
needs to be escaped (it's picked up as a H1 heading otherwise), but feel free to spell it "Number" if that looks prettier to you :)

```suggestion
	uint32_t delay;                 /**< \# of intervals to delay */
```

peter-mitsis replied:
Fixing.

kartben commented on samples/subsys/lw_sched/README.rst:
```diff
@@ -0,0 +1,59 @@
+.. zephyr:code-sample:: lw_sched
+   :name: Lightweight (LW) Scheduler
```
nit: unless it's like a proper name, or a well-known acronym i'd like to keep the sample names non capitalized. Looks like you're mostly spelling it in lower case yourself anyway!

```suggestion
   :name: Lightweight (LW) scheduler
```

peter-mitsis replied:
Fixing.

kartben reviewed (changes requested):
Trying to refrain from swearing in a public place but *****  :D As discussed on Discord, please see what I hope if my full review. Most of my comments didn't make it earlier :(

peter-mitsis commented on subsys/lw_sched/Kconfig:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config LW_SCHEDULER
+	bool "Light Weight Scheduler"
+	default n
+	help
+	  This option enables the use of the light weight schedulers. Each
+	  active light weight scheduler is a dedicated thread that runs at
+	  regular intervals to execute a list of tasks in sequence.
```
Fixing.

peter-mitsis commented on doc/services/lw_sched/index.rst:
```diff
@@ -0,0 +1,194 @@
+Delaying LW Tasks
+=================
+
+LW tasks can be arbitrarily delayed by invoking :c:func:`ll_task_delay`.
```
Fixing.

peter-mitsis commented on doc/services/lw_sched/index.rst:
```diff
@@ -0,0 +1,194 @@
+
+    struct task1_args execute_args1;
+
+    struct lw_task_args = {
```
Fixing

peter-mitsis commented on subsys/CMakeLists.txt:
Adding.

peter-mitsis commented on doc/services/lw_sched/index.rst:
```diff
@@ -0,0 +1,194 @@
+the order in which the LW tasks execute, wherein LW tasks with a lower 'order'
+are executed before those with a higher 'order'.
+
+LW tasks are initially in the paused state. They are subseqently moved into
```
Fixing.

peter-mitsis commented on doc/services/lw_sched/index.rst:
```diff
@@ -0,0 +1,194 @@
+
+It is important to understand that the LW scheduler is built atop the Zephyr
+scheduler and care must be taken with regards to blocking calls. Blocking calls
+not only block the LW task, but also that LW scheduler AND all of its tasks.
```
Fixing.

peter-mitsis commented on samples/subsys/lw_sched/README.rst:
```diff
@@ -0,0 +1,59 @@
+
+Overview
+********
+This sample demonstrates the use of the LW scheduler, and goes through the
```
Fixed.

peter-mitsis commented on samples/subsys/lw_sched/README.rst:
```diff
@@ -0,0 +1,59 @@
+This sample demonstrates the use of the LW scheduler, and goes through the
+following steps:
+
+   #. Initializes the LW scheduler that executes every 50 msec
+   #. Starts two sample tasks that execute every interval
+   #. Display the tasks' count info every second for 10 seconds
+   #. Abort the LW scheduler
+   #. Display the tasks' final count info
```
Fixed

martinjaeger commented on subsys/lw_sched/lw_scheduler.c:
```diff
@@ -0,0 +1,176 @@
+					break;
+				}
+
+				/* Fall through to the abort case */
```
```suggestion
				__fallthrough;
```

peter-mitsis replied:
I was not previously aware of this attribute. Now I am. :)

martinjaeger commented on include/zephyr/lw_sched/lw_sched.h:
```diff
@@ -0,0 +1,221 @@
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
+ */
+
+#ifndef __SUBSYS_LIGHTWEIGHT_SCHEDULER
```
Please use usual Zephyr header guard conventions:

```suggestion
#ifndef ZEPHYR_INCLUDE_LW_SCHED_LW_SCHED_H_
```

peter-mitsis replied:
Shall do.

martinjaeger reviewed (dismissed):
I'm wondering if we could achieve a very similar result without introducing a new subsystem by extending the delayable work queues with a feature to automatically reschedule a work item in a regular interval? This would even have the advantage that the period can be different for any work item.

When I read "lightweight scheduler" I thought this would be something more efficient than the regular scheduler, requiring less processing effort. But is it actually light-weight if it sits on top of the normal scheduler? The advantage compared to regular threads is that you don't need an individual stack space for each thread, right? But isn't hat exactly what work queues are for?

peter-mitsis commented on subsys/lw_sched/lw_scheduler.c:
```diff
@@ -0,0 +1,176 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
```
Good catch!

peter-mitsis commented:
I think I have addressed the most significant issues identified thus far.

andyross commented on subsys/lw_sched/lw_scheduler.c:
```diff
@@ -0,0 +1,176 @@
+	 * task from the list of tasks and insert it at the correct priority.
+	 */
+
+	while ((reorder = sys_dlist_get(&sched->reorder)) != NULL) {
```
This makes insertion/removal O(N), which makes complicated priority management expensive for large task lists.  Given that the whole point to this is to be "light weight" and suitable for using with hundreds/thousands of tasks, that's probably something we want to fix.

Use a rbtree (or some other log-or-better-time collection, but we already have an rbtree) instead?  Doesn't look like it should be difficult to port.

peter-mitsis replied:
Working on it.

peter-mitsis replied:
Now I remember why I went with the doubly linked list--despite the the O(N) cost ... removing nodes from an RB tree while it is being traversed can lead to nodes being visited twice or skipped, so I went with simplicity.

andyross commented on subsys/lw_sched/lw_scheduler.c:
```diff
@@ -0,0 +1,176 @@
+
+				__fallthrough;
+
+			case LW_TASK_ABORT:
```
What's the point to leaving a task in the list in an "ABORT" state, exactly?  Is it just to handle the race against the callback?  Seems like this might be a hole, if user code wants to be able to reuse a task but the list won't let it go yet?

peter-mitsis replied:
Once the LW scheduler detects that a task is in the abort state, it gets removed from the list--not much point in changing the state after this. If a developer wants to re-use the task, it must be reinitialized with lw_task_init() and started with lw_task_start().

andyross reviewed (commented):
A few notes after a brief read.  I will say I like this a whole lot better than the SOF code :)

kartben reviewed (dismissed):
Great, thanks for all the efforts on incorporating the feedback related to the documentation! +1 from me on that front.

peter-mitsis commented:
Updated to resolve a merge conflict.

ceolin reviewed (dismissed):
It is looking good :)

pdgendt commented on subsys/lw_sched/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if (CONFIG_LW_SCHEDULER)
```
This isn't needed, as this directory is added with `add_subdirectory_ifdef(CONFIG_LW_SCHEDULER lw_sched)`

peter-mitsis replied:
Removing ...

pdgendt commented on subsys/lw_sched/Kconfig:
```diff
@@ -0,0 +1,11 @@
+
+config LW_SCHEDULER
+	bool "Lightweight Scheduler"
+	default n
```
Remove `default n`

peter-mitsis replied:
Dropping ...

pdgendt commented on tests/subsys/lw_sched/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_ZTEST=y
+CONFIG_ZTEST_NEW_API=y
```
No longer needed (if rebased on main)

peter-mitsis replied:
Rebasing and removing CONFIG_ZTEST_NEW_API

peter-mitsis commented on subsys/lw_sched/lw_task.c:
```diff
@@ -0,0 +1,174 @@
+{
+	k_spinlock_key_t  key;
+
+	if ((task == NULL) || (ops == NULL) || (args == NULL) ||
```
Seems reasonable.  Making the change ...

peter-mitsis commented:
Updated to address comments from both @ceolin and @pdgendt .

carlescufi commented on doc/services/lw_sched/index.rst:
```diff
@@ -0,0 +1,195 @@
+    :local:
+    :depth:  2
+
+The lightweight (LW) scheduler is a dedicated thread that runs at regular
```
Could you please add a section here as to why one would use this instead of a workqueue? Coming from with no background knowledge whatsoever this is the first question that popped into my mind.

henrikbrixandersen replied:
On a related note: Please follow up on the questions from @martinjaeger regarding expanding the workqueue functionality with periodic calls.

fabiobaltieri commented on subsys/lw_sched/lw_scheduler.c:
```diff
@@ -0,0 +1,179 @@
+
+		/* Wait until the timer gives the semaphore */
+
+		k_sem_take(&sched->sem, K_FOREVER);
```
I think you could ditch the semaphore and synchronize the thread to the timer directly with `k_timer_status_sync`.

peter-mitsis commented:
### Differences to work queues

There are admittedly many similarities between the LW scheduler and work queues. There are several key differences with this current iteration.

**Ordering.**
Work queues are FIFO based--first one to be submitted to the work queue is the first one to be executed. LW tasks on the other hand have an _order_ field, much like a thread's priority. That is, they can be added in any order, but will execute according to that _order_ field (lowest to highest).

**Processing**
Work queues will process until all the work has been consumed in the order that they were submitted. The LW scheduler processes until it has completed a single pass of its work items, with the expectation that it will begin to re-process that list at the start of the next period/interval.

**Arguments**
Unless I have missed it, the work queue handlers does not explicitly have any custom arguments; however LW tasks do. That being said, this is relatively minor, as I expect that work queue handlers can obtain any necessary arguments from a parent structure that encapsulates the k_work structure the work queue handler uses (see below). So let's just consider this one  irrelevant.
```
    struct parent {
        struct k_work work;
        ... extra args ...
   };

```
Let's look at the question "Can work queues be used to perform this functionality?"

The answer to that question is _yes with a but_, The _but_ is that the developer must implement some extras and follow some guidelines.

Of the two remaining key differences, the _processing_ one is easiest to eliminate. One of the work items would need to be a synchronisation item to repeatedly sync with the trigger source (such as a k_timer) which would be created and started by the developer. This synchronising work item would need to be re-submitted to the work queue by the final work item to ensure that it blocks waiting for the semaphore to be given by the trigger source. It is a little extra work for the developer, but likely in the realm of reasonable.

This leaves _ordering_. The developer could maintain an ordered list of work items that ought to be processed this "period" of the work queue. The synchronising work item would before it is done, submit the next item in this external list to be the next work item processed by the work queue. the entire list has been traversed, then the next work item is the synchronising work item (it could even be set up as a sentinel). Maintenance of this external list would be up to the developer. As submission to the work queue would be done "just in time", there would be a greater overhead between calling work items than there would be for LW tasks. This might not be an unreasonable solution.

peter-mitsis commented:
[Lightweight Scheduler.pptx](https://github.com/zephyrproject-rtos/zephyr/files/13350052/Lightweight.Scheduler.pptx)

Attaching slides.
